journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+package com.evolutiongaming.kafka.journal.util
+
+import cats.instances.all._
+import cats.effect.implicits._
+import cats.effect.kernel.{Async, Deferred, Sync}
+import cats.effect.std.Mutex
+import cats.effect.{Ref, Resource}
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.Log
+
+import java.util.UUID
+import java.util.concurrent.TimeoutException
+import scala.concurrent.duration.FiniteDuration
+
+private[journal] trait ResourcePool[F[_], O] {
+
+  /**
+   * `acquire` function of the Resource semantically blocks if there are no objects available.
+   *
+   * `release` function of the Resource returns the object to the pool.
+   * Note that it will NOT deallocate the object.
+   */
+  def borrow: Resource[F, O]
+}
+
+private[journal] object ResourcePool {
+
+  type ObjectId = UUID
+  type WaiterId = UUID
+  type Waiter[F[_], O] = Deferred[F, Either[Throwable, (ObjectId, O)]]
+
+  case class AllocatedObject[F[_], O](value: O, deallocate: F[Unit], state: AllocatedObject.State)
+
+  object AllocatedObject {
+    sealed abstract class State(val isFree: Boolean)
+
+    object State {
+      case object Used extends State(false)
+      case class Free(releasedAt: FiniteDuration) extends State(true)
+    }
+  }
+
+  case class State[F[_], O](
+    objects: Map[ObjectId, AllocatedObject[F, O]],
+    waiters: Map[WaiterId, Waiter[F, O]],
+  ) {
+
+    def findFreeObject: Option[(ObjectId, AllocatedObject[F, O])] =
+      objects.collectFirst {
```

Comment:
I don't like performance characteristics of `findFreeObject`, it has a linear complexity. 
In often / worst case we will iterate through let's assume ~1k of entries to just find out that there is not a single one free…

Reply:
why do you think so?

Reply:
I think `scale = 0.5` is just a value we arrived to by trial and error :)

But in any case, you can tweak parameters in `ResourcePoolSpec` and see that the test completes very fast. I tried these params
```
val poolSize = 1000
val sleepDuration = 1.nanosecond
val timesToAcquire = 100_000
val idleTimeout = 1.second
val acquireTimeout = 30.seconds
```
and the test completed in a couple of seconds. So iterating through 1k elements in a map is not a bottleneck.